Some progress in Phys excph

rreho opened the pull request:
This PR builds upon the work initiated by @muralidhar-nalabothula in https://github.com/yambo-code/yambopy/pull/57, and introduces several new features and improvements related to exciton-phonon interactions and phonon-assisted photoluminescence.

# Summary of Changes

- New post-processing module: Added a dedicated folder and object for computing phonon-assisted photoluminescence (PL) based on optical property outputs.

- Exciton-phonon coupling: Implemented optimized routines to compute exciton-phonon coupling within yambopy.

- I/O improvements: Fixed data-type inconsistencies and improved reading/writing for several existing classes, notably LetzElPhCode and excitonDB, ensuring precision consistency across modules.

# Documentation:

A draft tutorial notebook, `exciton_phonon.ipynb`, showcasing the new functionality is available. 

This notebook can be built as part of the Qurex-Book using make book inside the docs folder (see [PR #46](https://github.com/yambo-code/yambopy/pull/46)).

The book is a work in progress but already includes this new example on phonon-assisted PL for hBN.

- `AUTHORS.md`: Updated my affiliation.

# Next Steps Before Merging
I would like to:

Discuss potential improvements to the structure of folders and classes in yambopy, aiming for cleaner modularity and long-term maintainability.

Perform additional testing with users and various datasets.

Add features related to exciton symmetries.

Improve the `exciton_phonon.ipynb` tutorial, which is currently a preliminary draft.

# Note!!!!
- This PR is built on top of an open PR https://github.com/yambo-code/yambopy/pull/68 . Ideally, first `fix-eps2` should be merged, then this one.
-The interface to Wannier90 is not yet included in this PR. I am holding off on pushing it until further development, but I hope to integrate it in the near future.

muralidhar-nalabothula commented on yambopy/optical_properties/ex_phonon.py:
```diff
@@ -0,0 +1,239 @@
+        nbnds = max(bands_range)-min(bands_range)
+        start_bnd_idx = 0
+        end_bnd = start_bnd_idx + nbnds
+        Dmat_data = Dataset(f'{self.LELPH_dir}/ndb.Dmats', 'r')
+        Dmats = Dmat_data['Dmats'][:, :, 0, start_bnd_idx:end_bnd,
+                                    start_bnd_idx:end_bnd, :].data
+        Dmats = Dmats[...,0] + 1j * Dmats[..., 1]
+        Dmats = Dmats.astype(self.wfdb.wf.dtype)
+        self.Dmats    = Dmats #self.wfdb.save_Dmat
+        Dmat_data.close()
```
replace this entire block with.
```python
self.Dmats = self.wfdb.Dmat()
```

rreho commented:
Hi,
Those lines were there due to the small differences I spotted few days ago and I forgot to replace it.
Now, I updated it with
`self.Dmats = self.wfdb.Dmat()[:,:,0,:,:]`  

Note the 3rd dimension is the spin dimension.

I am also working on generalizing the computation of `excPh` for finite Q. 
It seems to me that I have to change slightly `exciton_X_matelem`. The reason is that now I need $A_{k+q,cc}^{S Q+q}$ in addition to $A_{k,cc'}^{S'Q+q}$.

I'll have a look at it again and come back to this issue. In the meantime, I was wondering if we have a ref for tha values of exc-phonon coupling G at finite Q @muralidhar-nalabothula

rreho commented:
# Update 

-**Without** editing the method `exciton_X_matelem` (aside from adding few comments), I managed to compute the exciton-phonon coupling matrix elements at different Q. I have also cleaned up the class dividing the complex code in several small internal class methods. If I am not mistaked, the method relies on a translation from k-> k-q for $A_{k+q,cc}^{S Q+q}$  which becomes effectively $A_{k,cc}^{S Q+q}$. Hope I implemented it correctly 

- I have enabled efficient caching related to the IO of electron-phonon databases and the call to the `rotate_exc_wavefunction` class.

- ex-ph coupling can be computed in gamma only mode 

Results for luminescence at Q=0 remains ok for hBN bulk.


**TODO**  
- Test implementation at finite Q testing with Yambo code.
- Discuss correct usage of `rotate_exc_wavefunction`

rreho commented on yambopy/optical_properties/ex_phonon.py:
```diff
@@ -163,134 +169,195 @@ def read_excdb(self, BSE_dir):
+        # Original shape: (nb1, nb2, 2 spins, nm, nk) → we keep only spin 0 and swap bands for compatibility
+        return eph_mat_iq[:, :, 0, :, :].transpose(1, 0, 3, 2)
+    
+    def _rotate_exciton_pair(self, iq, iQ):
```
Danger zone! Check that proper arguments are passed to _rotate_exciton_wavefunction. The goal is to compute the rotated $A_{kcv}^{S,Q}$ and $A_{kcv}^{S,Q+q}$.

Note that right now I loop over Q points inside the IBZ.

palful commented:
I am checking the PR anyway, it will take some time because it contains many new things. Please stand by until I have tried all the new features, I will let you know my comments.

rreho commented:
Thanks Fulvio! I will split it in multiple PRs. I'll just wait a bit for your comments and to finalize other projects.

palful commented:
Hi Riccardo, I attach a .tar.gz file with all the inputs and scripts I used!
[testsPR70.tar.gz](https://github.com/user-attachments/files/22343159/testsPR70.tar.gz)

muralidhar-nalabothula commented:
Hi Fulvio,

Thank you very much for the review. Firstly, I think we should break this PR as this is too large to review. I will reopen my old #57 PR and we can first merge it and later slowly add exph after testing. I also like Riccardo's docs, but better to separate it out.

###  Dependencies :

- Regarding the dependencies, both are optional, but highly recommended (```pykdtree``` is 2-3 times faster.) ```Numba``` is a nice way to speedup.

###  Degeneracy_finder
- In ```degeneracy_finder```, the loop is there to take care of cases where consecutive values are with in tolerence. for example, consider this array ```array([1.   , 1.001, 1.002, 1.003, 1.004, 1.005, 1.006, 1.007, 1.008, 1.009])```, do we treat all of them as generate within ```0.001``` threshold?. This is debatable, for now I will remove the loop as such cases might be rare near the bandgaps.

### Exciton spin
1) Wrong doc string for ```compute_exc_spin_iqpt```. Updated. 
2) S_z is hermitian, so should be real, but I would leave it complex.
3) S^2 will always be zeros as the electron and hole "average" cancel each other (This can be seen directly by setting $g_{cc'/vv'}$ in exciton phonon matrix elements to $\delta_{vv'/cc'}$. This implies expectation value of S^2 will not tell you if it is single or triplet.

### Exciton wave functions in real space
- Davide changed ```Q-point``` to ```BS_Q``` in lumen i guess. Maybe we should request him to revert this change.


Regarding the exph implementation, I never had enough time to check/test Riccardo's implementation. I was expecting that they will work as they were directly taken from my PhD scripts repo and modified to use the new yambopy functionality such as wfdb class, new exciton functions etc. Something is not done correctly in that class. I checked with my old PhDscripts, I get similar to yambo figure. Nevertheless, we can also implement the entire exph quickly with the barebones in few lines like the below.
```python
# Read lattice
lattice = YamboLatticeDB.from_db_file(filename=ns_db1)
# Read electron-phonon
elph    = LetzElphElectronPhononDB(ndb_elph,read_all=False)
# Read wave functions
wfcs    = YamboWFDB(filename='ns.wf',save=savepath,latdb=lattice,bands_range=bands_range)

## load exe dbs
exdbs = []
for ik in range(wfcs.nkpoints):
    filename = 'ndb.BS_diago_Q%d' % (ik+1)
    excdb = YamboExcitonDB.from_db_file(lattice, filename=filename,
                                             folder=bsepath,
                                             Load_WF=True, neigs=10)
    exdbs.append(excdb)


exph_mat = []
Dmats = wfcs.Dmat()
# compute ex-ph
for iq in range(elph.nq):
    ph_eig, elph_mat = elph.read_iq(iq,convention='standard')
    elph_mat = elph_mat.transpose(1,0,2,4,3)
    #
    idx_BZq = wfcs.kptBZidx(elph.qpoints[iq])
    iq_isymm = lattice.symmetry_indexes[idx_BZq]
    iq_iBZ = lattice.kpoints_indexes[idx_BZq]
    trev  = (iq_isymm >= len(lattice.sym_car) / (1 + int(np.rint(lattice.time_rev))))
    symm_mat_red = lattice.lat@lattice.sym_car[iq_isymm]@np.linalg.inv(lattice.lat)
    #
    exe_iqvec = wfcs.kpts_iBZ[iq_iBZ]#exdbs[iq_iBZ].car_qpoint
    #
    Akq = rotate_exc_wf(exdbs[iq_iBZ].get_Akcv(),symm_mat_red,wfcs.kBZ,exe_iqvec,Dmats[iq_isymm],trev,wfcs.ktree)
    #
    tmp_exph = 0.5 * exciton_X_matelem(np.array([0,0,0]), elph.qpoints[iq], Akq, exdbs[0].get_Akcv(), elph_mat,
                      wfcs.kBZ, contribution='b', diagonal_only=False, ktree=wfcs.ktree)
    ## 0.5 for Ry to Ha.
    exph_mat.append(tmp_exph)


exph_mat = np.array(exph_mat).transpose(0,1,3,2) #(q,nmmodes, initial (Gamma), final (Q))
np.save('Ex-ph.npy',exph_mat)

```
The above code is consistant with PhD scripts and is gauge invariant. It also does not depend on what convention we use for letz code. Below is what you get from the above script.


<img width="640" height="480" alt="Figure_1xxxxx" src="https://github.com/user-attachments/assets/42e2a7ec-538a-4c85-aaa7-557eda11ecc8" />


So I suggest, we start adding things slowly.



ps. I tried the yambo_ph input, but I got bad values, not sure why (I used lumen ex-ph branch). If possible  Could you please see if the above scripts exactly same figure?

rreho commented:
<img width="568" height="442" alt="test-kplot" src="https://github.com/user-attachments/assets/bfab4d60-0ade-4376-863a-4171205f2d7a" />

I believe I fixed the bug relative to the k-space plot. Works both in the gamma only case or without.
Passing by, I noted that the way I treat the various energy conversion in `LetzElphElectronPhononDB` diverged w.r.t to MN's forked yambopy. I tested that both give the same final results, but small numerical discrepancies are there.

I still don't get why you include a 0.5 factor. I do not. Moreover, it appears that the absolute value displayed on the `cmap` is not exactly equal. Anyone has an idea why the magnitude would be slightly different? I also noted that the raw data values are all quite small and around 1e-5/1e-6. Clearly, in the final result we sum, but still, perhaps small discrepancies are expected.

palful commented:
Thank you Riccardo! Let me know when you finished the cleaning and please re-synchronize this with `yambo-code:bug-fixes` which now includes PR #57. Then I will go back to testing the small discrepancies.

By the way: what do you mean when you say that there are discrepancies in energy conversion in the letzelphc db class? And which 0.5 factor are you referring to?

palful commented:
Hi Riccardo, thanks a lot! Thanks also for your other comment on the lumen PR and checking the tutorial. I'm working on reconfiguring the IO of the dipoles part, I will let you know!

palful commented:
Ok, so I updated `YamboDipolesDB` in the bug-fixes to accept `bands_range` but it was a real mess. It needed to have compatibility for the `DipBandsAll` case, with `YamboElectronsDB`, dipole expansion, k-plots and IP absorption... It took some time to make sure everything worked. 

In the process, I changed the class to use the `from_db_file` method to load the database. Although this makes the update not compatible with users' scripts, on the other hand it made the changes easier and gave the class the same loading method as other classes, avoiding confusion. Now the class can be called as `YamboDipolesDB.from_db_file(latdb,filename=f'{dipoles_path}/ndb.dipoles',bands_range=bands_range)`. At the end of this I will make a new release where this change is noted.

Can you please try to merge this into the PR? Afterwards, I will check the rest!

rreho commented:
Impressive work @palful. I was expecting the dipoles to be a nightmare and it required several important decisions on the repo. Thanks!

I merged, adapted my branch to your changes and tested (only) that I reproduce the excph plot above. I also add to change slightly the import orders bsekerneldb.py and dipolesdb.py which I believe was a bug present in `bug-fixes`
